### PR TITLE
fix #291 + #293 + #313 : Session.save (0D arrays + Excel + overwrite file by default) 

### LINKED ISSUE #291: Session.save fails when it contains 0d arrays
because `LArray(0).to_frame()` fails

### PR DESCRIPTION
**Pull request by alixdamman**
until #153 is fixed.

### CODE REVIEW

File: larray/io/session.py [hunk 1]
```diff
@@ -114,6 +115,10 @@ def dump_arrays(self, key_values, *args, **kwargs):
         display = kwargs.pop('display', False)
         self._open_for_write()
         for key, value in key_values:
+            if isinstance(value, ABCLArray) and value.ndim == 0:
+                if display:
+                    print('Cannot dump {}. Dumping 0D arrays is not supported currently.'.format(key))
```
Comment (gdementen) [inline larray/io/session.py]: currently not supported

**Comment by alixdamman** [comment]
Did you run pytest after commit https://github.com/liam2/larray/pull/309/commits/1286d6715fb7244f60762c2d75c86078319e1064 ? Because you removed the "__larray__" method from Range class but the call from "__getattr__" a dew lines below.

**Comment by gdementen** [comment]
Oops, nice catch! I thought I did run the tests. Maybe I will need to add another one.

File: doc/source/changes/version_0_24.rst.inc [hunk 2]
```diff
@@ -57,3 +57,9 @@ Fixes
 * fixed the array() method on excel.Sheet returning float labels when int labels are expected.
 
 * fixed getting float data instead of int when converting an Excel Sheet or Range to an larray or numpy array.
+
+* fixed crash of Session.save method when it contains a 0D array.
+  0D arrays are skiped when saving a session (closes issue:`291`).
+
+* fixed crash of Session.save method when trying to save a session in an Excel that does not exist.
+  A new Excel file is created if it does not exist yet (closes issue:`313`).
```
Comment (gdementen) [inline doc/source/changes/version_0_24.rst.inc]: isn't crash a too strong word for something which failed with a relatively nice error message? I would write something like:

    fixed Session.save and Session.to_excel failing to create new Excel files (it only worked if the file already existed). Closes ...

File: larray/tests/test_session.py [hunk 3]
```diff
@@ -161,6 +162,8 @@ def test_xlsx_pandas_io(self):
     @pytest.mark.skipif(xw is None, reason="xlwings is not available")
     def test_xlsx_xlwings_io(self):
         fpath = abspath('test_session_xw.xlsx')
+        if os.path.isfile(fpath):
```
Comment (gdementen) [inline larray/tests/test_session.py]: I think we also need an explicit test where the file already existed.

**Comment by alixdamman** [comment]
Do you want me to include issue #293 in this PR or another?

**Comment by gdementen** [comment]
as you prefer.

**Comment by alixdamman** [comment]
Not sure for issue #293

File: larray/tests/test_session.py [hunk 4]
```diff
@@ -134,24 +136,39 @@ def test_names(self):
...
         s = Session()
         s.load(fpath, engine='pandas_excel')
         self.assertEqual(list(s.keys()), ['e', 'g', 'f'])
 
+        # update an array (overwrite=False)
+        s['e'] = self.e2
+        s.save(fpath, engine='pandas_excel', overwrite=False)
+        s.load(fpath, engine='pandas_excel')
+        self.assertEqual(list(s.keys()), ['e', 'g', 'f'])
+        #assert_array_nan_equal(s['e'], self.e2)
```
Comment (gdementen) [inline larray/tests/test_session.py]: why is this line commented?

File: larray/tests/test_session.py [hunk 5]
```diff
@@ -161,12 +178,21 @@ def test_xlsx_pandas_io(self):
     @pytest.mark.skipif(xw is None, reason="xlwings is not available")
     def test_xlsx_xlwings_io(self):
         fpath = abspath('test_session_xw.xlsx')
+        # test save when Excel does not exist
```
Comment (gdementen) [inline larray/tests/test_session.py]: Excel file

File: larray/tests/test_session.py [hunk 6]
```diff
@@ -134,24 +136,39 @@ def test_names(self):
...
+
         s = Session()
         s.load(fpath)
         # HDF does *not* keep ordering (ie, keys are always sorted)
         self.assertEqual(list(s.keys()), ['e', 'f', 'g'])
 
+        # update an array (overwrite=False)
+        s['e'] = self.e2
+        s.save(fpath, overwrite=False)
+        s.load(fpath)
```
Comment (gdementen) [inline larray/tests/test_session.py]: This test (and all similar tests below) is a bit misleading. It will pass even if the file is completely wiped out, because the arrays still exist in the session s and s.load(path) does not clear the session. Maybe we want to change that last point, but that would be for another PR/release.

File: larray/core/session.py [hunk 7]
```diff
@@ -245,12 +245,14 @@ def save(self, fname, names=None, engine='auto', display=False, **kwargs):
         names : list of str or None, optional
             List of names of objects to dump. If `fname` is None, list of paths to CSV files.
             Defaults to all objects present in the Session.
-        engine : str, optional
+        engine : {'auto', 'pandas_csv', 'pandas_hdf', 'xlwings_excel', 'pickle'}, optional
             Dump using `engine`. Defaults to 'auto' (use default engine for
             the format guessed from the file extension).
+        overwrite: bool, optional
+            Whether or not to overwrite an existing file, if any. Only for non CSV files.
```
Comment (gdementen) [inline larray/core/session.py]: Ignored for CSV files?

File: doc/source/changes/version_0_24.rst.inc [hunk 8]
```diff
@@ -18,6 +18,8 @@
      a0   0   1
      a1   2   3
 
+* added new boolean argument 'overwrite' to Session.save method. If True, overwrite existing file if any.
```
Comment (gdementen) [inline doc/source/changes/version_0_24.rst.inc]: * we should also mention other "save" methods: to_excel, etc. which are also impacted by this change.
* I don't think it is clear enough that the default behaviour changed from overwrite=False to overwrite=True

File: larray/io/session.py [hunk 9]
```diff
@@ -164,7 +169,7 @@ def list(self):
 
     def _read_array(self, key, *args, **kwargs):
         df = self.handle.parse(key, *args, **kwargs)
-        return df_aslarray(df)
+        return df_aslarray(df, raw=True)
```
Comment (gdementen) [inline larray/io/session.py]: What's the relation with the rest? Is this an extra fix? If so it needs a test & changelog :)

Comment (alixdamman) [inline larray/io/session.py]: This is an extra fix for an issue discovered when I tried to update the IO tests for Session ;)

Comment (gdementen) [inline larray/io/session.py]: In fact, this is not entirely correct. We should use: raw=index_col is None. But I admit the chance users ever hit this case is close to zero :)

Comment (alixdamman) [inline larray/io/session.py]: forgot changelog, sorry.

**Comment by gdementen** [comment]
LGTM
